keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -526,302 +517,382 @@ dualstack cluster can have this flag configured as
+  is not configured for dual stack) then they can do one of the following:
+  1. Either don't set `spec.preferDualStack` or set it to `true`, and set
+     `spec.ipFamilies` to the two IP families in any order.
+  2. Either don't set `spec.preferDualStack` or set it to `true`, and set
+     `spec.clusterIPs` to any two IPs of different families, in any order.
+
+The below are sample services that demonstrate the above rules.
+
+Given a cluster configured for dual stack as `<ipv4-cidr>,<ipv6-cidr>`, this
+input:
 
 ```yaml
-# dual stack service (required)
 apiVersion: v1
 kind: Service
 metadata:
   name: my-service
 spec:
   type: ClusterIP
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
   preferDualStack: true # set by user
-  ipFamilies: # set by user
+  ipFamilies:           # set by user
     - IPv4
     - IPv6
-  clusterIP: 1.2.3.4  # allocated
-  clusterIPs: # allocated
-    - 1.2.3.4
-    - 2001::1
+```
+
+...will produce this result:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
   selector:
     app: MyApp
   ports:
     - protocol: TCP
       port: 80
       targetPort: 9376
+  preferDualStack: true # set by user
+  ipFamilies:           # set by user
+    - IPv4
+    - IPv6
+  clusterIP: 1.2.3.4    # allocated
+  clusterIPs:           # allocated
+    - 1.2.3.4
+    - 2001::1
+```
+
+If the cluster had not been configured for dual stack, this would have failed.
+The user could have specified `ipFamilies` in the alternate order, and the
+allocated `clusterIPs` would have switched, too.
+
+This input:
 
-# dual stack service (required)
+```yaml
 apiVersion: v1
 kind: Service
 metadata:
   name: my-service
 spec:
   type: ClusterIP
-  preferDualStack: true # defaulted
-  ipFamilies: # defaulted
-    - IPv6
-    - IPv4
-  clusterIP: 2001::1  # set by user
-  clusterIPs: # set by user
-    - 2001::1
-    - 1.2.3.4
   selector:
     app: MyApp
   ports:
     - protocol: TCP
       port: 80
       targetPort: 9376
+  preferDualStack: true # set by user
+```
 
-# dual stack service (optional)
+...would produce the following result on a single stack cluster:
+
+```yaml
 apiVersion: v1
 kind: Service
 metadata:
   name: my-service
 spec:
   type: ClusterIP
-  preferDualStack: true # set by user
-  ipFamilies:
-    - IPv6 # defaulted, cluster does not support dual stack services, only one IPv6 CIDR is provided for services
-  clusterIP: 2001::1  # allocated
-  clusterIPs: # allocated
-    - 2001::1 # note single entry
   selector:
     app: MyApp
   ports:
     - protocol: TCP
       port: 80
       targetPort: 9376
+  preferDualStack: true # set by user
+  ipFamilies:           # defaulted
+    - IPv6              # note single entry
+  clusterIP: 2001::1    # allocated
+  clusterIPs:           # allocated
+    - 2001::1           # note single entry
+```
+
+On a dual stack cluster it would have produced:
```

Comment:
"On a dual-stack cluster with a primary cluster IP family of IPv6 it would have produced"